### DISCUSSION TITLE
GLFW questions

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#31] [Q&A]
Who is our resident GLFW expert? I have questions! :)

I am tweaking `minimal-ui.lisp` in the SBCL folder in the `merge-ccl-sbcl` branch.

Some questions:

1) When these callbacks are set, do the arguments have to be symbols or can they be functions? If they are symbols, the binding does not change when the function definition is changed. That means we have to reopen the window every time we make an edit to the callbacks. Since the callback functions are not regular CL functions, is this a limitation of working with GLFW?

```
(glfw:set-key-callback 'key-callback)
(glfw:set-mouse-button-callback 'mouse-callback)
(glfw:set-window-size-callback 'window-size-callback)
```

2) How do I get the results of a key and modifier together? For example, I want to test for "?", but I get "slash" and "left-shift".

3) I opened this issue for mouse-dragging: https://github.com/kaveh808/kons-9/issues/30

**Comment by kaveh808** [comment]
1) A workaround is to have the callback call a regular function, which does get updated when recompiled.

2) I guess I could test for :slash and the shift key.

3) Fixed and pushed. All (minimal) UI now working.

**Comment by awolven** [comment]
You have to "translate" the glfw key code into something lisp understands.

**Comment by awolven** [comment]
You have to define the callback with cffi:defcallback.

(defcallback imgui-glfw-key-callback :void ((user-data :pointer) (key :int)
(arg2 :int) (action :int) (mods :int))
  (declare (ignorable user-data arg2 mods))
  (imgui-glfw-key-callback-function key action))

Defcallback associates the symbol with a static c-accessible function
pointer.

In my glfw bindings, you then set the callback like this:

(glfwSetKeyCallback (h window) (callback imgui-glfw-key-callback))

with cl-glfw, it appears you just:

(glfw:set-key-callback 'key-callback)

Since it is inconvenient to call set-key-callback again after
recompiling, where the body of the callback has changed, you want to
"bounce" to a lisp function:

(defun imgui-glfw-key-callback-function (key action)

  (on-key-event *app* key action)

  (values))

You can then either bounce again to a method, or some kind of hook
which evaluates to some named function or a lambda.


Here is an example event method which should shed some light on the
key and the action (I'm not using the modifier arg):


(defmethod on-key-event ((app text-editor-mixin) key action)
  (reroute-debug
    (block on-key-event

      (when (or (eq key %glfw:GLFW_KEY_LEFT_SHIFT)
                (eq key %glfw:GLFW_KEY_RIGHT_SHIFT))
        (when (eq action %glfw:GLFW_PRESS)
          (setf (shift-key-down? app) t))
        (when (eq action %glfw:GLFW_RELEASE)
          (setf (shift-key-down? app) nil)))

      (when (or (eq key %glfw:GLFW_KEY_LEFT_CONTROL)
                (eq key %glfw:GLFW_KEY_RIGHT_CONTROL))
        (when (eq action %glfw:GLFW_PRESS)
          (setf (ctrl-key-down? app) t))
        (when (eq action %glfw:GLFW_RELEASE)
          (setf (ctrl-key-down? app) nil)))

      (when (or (eq key %glfw:GLFW_KEY_LEFT_ALT)
                (eq key %glfw:GLFW_KEY_RIGHT_ALT))
        (when (eq action %glfw:GLFW_PRESS)
          (setf (meta-key-down? app) t))
        (when (eq action %glfw:GLFW_RELEASE)
          (setf (meta-key-down? app) nil)))

      (when (or (eq key %glfw:GLFW_KEY_LEFT_SUPER)
                (eq key %glfw:GLFW_KEY_RIGHT_SUPER))
        (when (eq action %glfw:GLFW_PRESS)
          (setf (super-key-down? app) t))
        (when (eq action %glfw:GLFW_RELEASE)
          (setf (super-key-down? app) nil)))


      (unless (or (eq key %glfw:GLFW_KEY_LEFT_SHIFT)
                  (eq key %glfw:GLFW_KEY_RIGHT_SHIFT)
                  (eq key %glfw:GLFW_KEY_LEFT_CONTROL)
                  (eq key %glfw:GLFW_KEY_RIGHT_CONTROL)
                  (eq key %glfw:GLFW_KEY_LEFT_ALT)
                  (eq key %glfw:GLFW_KEY_RIGHT_ALT)
                  (eq key %glfw:GLFW_KEY_LEFT_SUPER)
                  (eq key %glfw:GLFW_KEY_RIGHT_SUPER))

        (when (or (eq action %glfw:GLFW_PRESS)
                  (eq action %glfw:GLFW_REPEAT))
          (multiple-value-bind (x y) (get-cursor-pos (main-window app))
            (hi::q-event hi::*editor-input*
                         (translate-glfw-key key (shift-key-down? app)
                                             (ctrl-key-down? app)
                                             (meta-key-down? app)
                                             nil)
                          (round x) (round y)))

          (return-from on-key-event (values)))))

    (values)))
